### PR TITLE
Outline10

### PR DESCRIPTION
**Pull request by speudusa**
The Methods Chapter.  bum ba ba! *jazz hands*

This was mostly from scratch, so any suggestions about overall flow, language, or complexity of code blocks are welcome.  I got myself stuck overthinking the studio.  Is this too simple?   I was hoping that this would fill in the basics of methods, so that when we are able to use them to instantiate objects we can focus more on access modifiers and calling between classes instead of worrying about signature and invocation 101 stuff.

### CODE REVIEW

**Comment by speudusa** [comment]
I hope i didn't add too much extra.  you've already looked at the loops...  i was in an merge flow after assignment 1.

File: src/chapters/methods/method-signatures.rst [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      static int SquareTheNumber(int num)
+      {
+         return num * num;
+      }
+
+
+In this example, we are using ``static`` as our **access modifier**.  This is lets C# know that this method
+belongs to the current class it is contained in and not to expect anything **new** to be created.  
+
+Our **return type** is **int**.  This lets you know how to use the method.  If you were expecting a **double** then you will needing
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-signatures.rst]: If you expect a **double**, then you will need* ...

File: src/chapters/methods/method-signatures.rst [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+``SquareTheNumber`` is the **method name**.  
+Notice that it is all in `PascalCase <https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions>`_ where all words are capitalized and spaces removed.
+
+``SquareTheNumber`` has a single **parameter**, ``int num``, which is a placeholder that tells you what **input** is expected along with the data type.  
+In this example, ``SquareTheNumber`` is expecting one input of type **int**.  The input is labeled ``num`` to provide you more *context* on what to expect or how to use the method.  
+
+When writing your own methods, provide as many clues as possible to help with readability.  
+The **method body** shows you how the input will be manipulated within the method.  
+Here it will be used to square itself.  This method will ``return`` a value which we can use else where in our code.  Since this method is static,
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-signatures.rst]: elsewhere is one word

File: src/chapters/methods/method-signatures.rst [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+   .. sourcecode:: bash
+
+      16
+      32
+      25
+
+
+You should be fairly familiar with the replit IDE by now.  
+
+- **Line 1** us where the ``using`` statement goes.
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-signatures.rst]: tells us?

File: src/chapters/methods/method-signatures.rst [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+      25
+
+
+You should be fairly familiar with the replit IDE by now.  
+
+- **Line 1** us where the ``using`` statement goes.  
+- **Line3** introduces the ``MainClass`` class followed by ``{``.  The closing ``}`` is in **Line 21**.  
+- Starting at **Line 11** is the Main method statement, followed by a ``{`` which closes on **Line 20**.
+
+We can place our method anywhere in the ``MainClass`` but outside the Main method.  In this example, we placed our
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-signatures.rst]: Im not crazy about the way this first statement is phrased. Can you rephrase it to be more explicit that this method can be anywhere in the class as long as it is not nested within the Main method's method body?

File: src/chapters/methods/method-signatures.rst [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+``SquareTheNumber`` method between **Lines 5-8**.  
+
+Calling the Method
+--------------------
+
+We are using static methods with expected returns, these take input and return output, or values.  So we must provide them with input if we want to see them work.  
+Since this method is returning a numeric value, it needs to be stored somewhere.  
+You can create a variable to do so, as we did in **Line 14** with the variable ``squaredNum``.
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-signatures.rst]: i think this is actually in line 13 now. And actually, this info could be more digestible by referencing the code snippet rather than the line #s - since the reader needs to scroll up or flip to the replit page to confirm what code exists on each line.

Check the line #s in the rest of this section as well.

Also, the repl linked here contains different code than what is discussed

File: src/chapters/methods/method-signatures.rst [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+      {
+         return (2 * x) + (2.5 * y);
+      }
+
+
+   #. The data type of the parameter is incorrect.
+   #. The method name should be ``perimeter``.
+   #. A return statement is inappropriate for this type of method.
+
+   .. ans: a.
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-signatures.rst]: Im not getting why a is the correct answer here :/

File: src/chapters/methods/using-methods.rst [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+------------------------------
+
+In the previous section, we looked at methods that return a value.  That value is then able to be used elsewhere in your program.  
+
+Sometimes we might want to use our method to manipulate data for the console output only. 
+This is when you may want to consider using a ``void`` return type.  
+
+.. admonition:: Example
+
+   .. sourcecode:: csharp
```
Comment (CarlyLanglois) [inline src/chapters/methods/using-methods.rst]: this example would do well with a repl linked as well so that the students have another opportunity to see how these two methods need to be defined within a class

File: src/chapters/methods/using-methods.rst [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+
+   :: 
+      
+      Willow's birthday is on 12/12
+      Lil's birthday is on 8/21
+      
+
+In this example, we are asking the method to print a friend's name and birth date inside a preprogrammed message.  
+This might see really simple, but if you had a list of names to run through, calling your method for each is a LOT less typing than 
+a ``Console.WriteLine`` statement with you message for each one.  That would get old fast.
```
Comment (CarlyLanglois) [inline src/chapters/methods/using-methods.rst]: your* message

File: src/chapters/methods/using-methods.rst [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+
+Inside the parentheses is ``Parameter List``.  In the example above ``BirthdayNotes`` has three items inside the parentheses.  
+These are the parameters the method is expecting.  Parameters are *placeholders* for the method while you build it and are optional.
+**Arguments** are the spcific, real values you pass to the method.  
+
+In the ``BirthdayNotes`` example, the parameter list shows us that the method is expecting a string value and two int values in that order.  
+For our human understanding we gave the parameters helpful names, but those names are NOT variables or values.  
+They are holding space for your actual arguments.  The method will take your arguments and use them as designated by the parameters.
+
+When we call the method, we can pass our arguements directly or via variables and check the output to see if we are happy with ther results.
```
Comment (CarlyLanglois) [inline src/chapters/methods/using-methods.rst]: the* results

Comment (CarlyLanglois) [inline src/chapters/methods/using-methods.rst]: arguments*

File: src/chapters/methods/using-methods.rst [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+
+   :: 
+      
+      Willow's birthday is on 12/12
+      This is also a string's birthday is on -55/987654321
+      
+
+In the example we called BirthdayNotes twice.  The first time, ``BirthdayNotes("Willow", 8, 3)`` and the second time ``BirthdayNotes("This is a string, too", 3-5, 987654321)``.  
+The code ran both times because the data types of the arguments matched those of the parameters.  The second call doesn't make any sense in the context
+of the method's overall function.  However, since it was provided with arguments of the same data type as the parameters, the method ran as expected.
```
Comment (CarlyLanglois) [inline src/chapters/methods/using-methods.rst]: maybe instead: "The second call doesn't make any sense in the context
of the method's overall job."

only to avoid using "function" as a description of what the method does

File: src/chapters/methods/using-methods.rst [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+
+         public static void Main (string[] args) {
+            
+         double purchase1 = 25.43;
+         double total1 = purchase1 + SalesTax(purchase1);
+         
+         Console.WriteLine(total1);
+         
+         }
+      }
```
Comment (CarlyLanglois) [inline src/chapters/methods/using-methods.rst]: can you un-indent this block so its on the same spacing as the method above?

File: src/chapters/methods/using-methods.rst [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+      
+      Carolyn worked at Tate & Bywater as a(n) Receptionist for 2 years.
+      Beth worked at Hope Center as a(n) RVT for 3 years.
+
+
+
+Optional Arguments
+^^^^^^^^^^^^^^^^^^^^
+
+An argument becomes optional when the parameters is coded to take a **default value**.
```
Comment (CarlyLanglois) [inline src/chapters/methods/using-methods.rst]: singular parameters

File: src/chapters/methods/method-recursion.rst [hunk 13]
```diff
@@ -0,0 +1,111 @@
+Recursive Methods
+=================================
+
+.. index:: ! recursion 
+
+If methods can use methods, can a method use itself?  
+
+The answer is yes.  This process is called **recursion** and is
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-recursion.rst]: make "this process..." into 2 sentences for clarity on the definition of recursion

File: src/chapters/methods/method-recursion.rst [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+.. admonition:: Example
+
+   .. sourcecode:: csharp
+      :linenos:
+      
+      using System;
+
+         class MainClass {
+         
+            static double FactorialRecursive(double number)
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-recursion.rst]: check your indentations here too

File: src/chapters/methods/method-recursion.rst [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+   **Console Output**
+
+   ::
+
+      5040 is the factorial outcome of 7
+
+
+Here is an example of a recursive method.  In **Lines 5-12** we create the function.  
+The base case is the ``if`` statement in **Line 7** which says that when the argument reaches ``0``, the method will return ``1`` and stop.
+But if the arguement is not ``0``, we leave the ``if`` statment and go to **Line 11** which uses the method to call itself once more, with modified parameters.
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-recursion.rst]: make this 2 sentences

File: src/chapters/methods/method-recursion.rst [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+.. admonition:: Example
+
+   .. sourcecode:: csharp
+      :linenos:
+      
+      using System;
+
+         class MainClass {
+         
+            static double FactorialLoop(double number)
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-recursion.rst]: indentation is off on this example too. maybe you should check them all :)

Comment (CarlyLanglois) [reply src/chapters/methods/method-recursion.rst]: this is a great example btw

File: src/chapters/methods/method-scope.rst [hunk 17]
```diff
@@ -0,0 +1,186 @@
+Scope 
+==========
+
+.. index:: ! scope
+
+**Scope** refers to visibility of a particular variable within your code.  
+So far we have defined variables directly in the Main method within the ``MainClass`` in the replit IDE, 
+but also inside loops and methods.  Perhaps you have exerienced an error when trying to print or call a variable that 
+was created inside you loop from outside of the loop.  This is an example of a scope issue.
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-scope.rst]: your* loop

File: src/chapters/methods/method-scope.rst [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+and can be stored inside a class or inside methods, loops, and conditionals.  
+Anything created within that code block will stay there unless returned or granted access.  
+Classes are larger boxes that can hold multiple loops and methods and, depending on the access modifiers, are able to access 
+all of them within it as well as items from other classes.  
+
+When you run into a scope error, the compiler error messages use the word **context** instead of scope.  
+When you see this message, you might need to backstep through your work and look at your boxes.  
+*Did you provide a way to open the lid?  Did you mean to work within the loop or method?  
+Did you really want* that *access modifier?  Did you accidently open the lid when it should stay closed?*
+Scope can vary from too narrow, as in trying to access a variable created within a loop or to broad,
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-scope.rst]: too* broad

Comment (CarlyLanglois) [reply src/chapters/methods/method-scope.rst]: also this sentence is too long. can you make it into 2 ... maybe 3

File: src/chapters/methods/method-scope.rst [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+         {
+            return num1+ num2+ num3;
+         }
+
+         public static void Main (string[] args) {
+
+            int total = SumItUp(1,2,3);
+            Console.WriteLine(total);
+         }
+      }
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-scope.rst]: ok this note is kind of nit picky but try keeping the bracket usage consistent. here we have the class and the main method with an opening bracket on the same line as the definition yet the SumItUp method's opening bracket is on the next line. 
I don't care which you use - although I think the unit 2 book uses almost exclusively opening brackets on the following line

File: src/chapters/methods/method-scope.rst [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+Scope can vary from too narrow, as in trying to access a variable created within a loop or to broad, 
+as in able to access data through the entire program which could cause issues with security or accidental data manipulation.
+
+In C#, scope is broken down into three categories: class level, method level, and block level.
+
+Class Level Scope 
+----------------------
+
+Class level scope you are already familiar with, but once we begin to add *more* classes to our code, this can become an issue. 
+When we get to classes, we will discuss how to handle class level scope in more detail. (Hint: think access modifiers).
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-scope.rst]: are they familiar with it yet? this may change depending where this chapter gets placed in lesson order

File: src/chapters/methods/method-scope.rst [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+In this example we are trying to print the variable ``total``.  
+We can access ``total`` because it has been defined within ``MainClass``, as you can see in **Line 1** in this example.
+If we had a second class, depending on the access modifier of the ``SumItUp`` method, we might not be able to access it.
+
+This will make more sense when we start learning about classes in future chapters.
+
+Method Level Scope 
+------------------------
+
+With method level scope, any variables declared **within** a method are not avaliable outside the method.
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-scope.rst]: within shouldn't be boldened here because it is not a vocabulary word. you can use just one set of asterisks like *this* to make the text italicized for emphasis however

File: src/chapters/methods/method-scope.rst [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+There are 3 examples of block scope in this example.
+
+Starting with **Line 8**, we print the value of ``i`` from inside the ``PrintThis`` method.  
+When this runs, you will see ``0,1,2`` print on new lines after calling the method in **Line 19**.
+It works because **Line 8** is able to see the values of ``i`` via the ``for`` loop code block.
+
+Moving down to **Line 14**, we see another ``for`` loop.  This loop will print each value of ``j`` from within its codeblock as is. 
+There are no scope issues here.  Run the code and you will see ``10, 13, 16, 19`` print to the console on new lines.
+
+**Line 20** is where we find a scope issue.  We want to know what ``j`` is **outside** the ``for`` loop.
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-scope.rst]: I would italics here again for the emphasis on *outside*

File: src/chapters/methods/method-scope.rst [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+However, ``j`` doesn't exist outside of the context of the ``for`` loop, so the ``Console.WriteLine`` statement in **Line 20**
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-scope.rst]: try adding: "However, ``j`` doesn't exist outside of the context of the ``for`` loop block ..."

to help make it more clear what a code block is

File: src/chapters/methods/method-scope.rst [hunk 24]
```diff
@@ -0,0 +1,186 @@
...
+won't be able to find any values and will throw an error.  In fact, the compiler won't run ANY of your code until you fix **Line 20**.
+
+Check Your Understanding
+-------------------------
+
+.. admonition:: Question
+
+   Identify the level of scope for each variable.  Bonus points if you can explain what will happen **Lines 16 & 17** and why.
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-scope.rst]: kind of annoying but I'd take out "bonus point for" and just ask them to consider this on their own. since we're not giving point for concept check questions, we would almost certainly get questions about this from students

File: src/chapters/methods/method-scope.rst [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+
+      Console.WriteLine(byHundred);
+      Console.WriteLine(byHalf);
+
+
+   #. Class scope
+   #. Method scope
+   #. Loop scope
+
+.. ans: 1 block, 2 method, 3 class
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-scope.rst]: a few issues with this question - a) the answer options dont match the answer :) and b) the container class isnt visible so a class level scope can be determined by deduction but really showing the container class would be a good fit here.

I'd also consider just asking: 'what is the scope level of the the variable `int byHalf`

File: src/chapters/methods/exercises.rst [hunk 26]
```diff
@@ -0,0 +1,188 @@
...
+   #########
+    #######
+     #####
+      ###
+       #
+
+.. admonition:: Note
+
+   As you build your methods, remember to keep the output in mind.  
+   Do you want a value returned or are do you want the method to print directly to the console when invoked?
```
Comment (CarlyLanglois) [inline src/chapters/methods/exercises.rst]: or are do you want - ?!

File: src/chapters/methods/exercises.rst [hunk 27]
```diff
@@ -0,0 +1,188 @@
...
+   How many things are being repeated?
+
+There is no starter code for these exercises, so create a new C# project
```
Comment (CarlyLanglois) [inline src/chapters/methods/exercises.rst]: normally, i'd support this approach but I think we should give them a starter repl with the container class and main method created since we want to focus of the practice to be on the method creation. what do you think?

File: src/chapters/methods/exercises.rst [hunk 28]
```diff
@@ -0,0 +1,188 @@
...
+
+      #####
+      #####
+      #####
+      #####
+      #####
+
+Asthetically it's not a "perfect" square, but a single argument led to its creation.
+
+#. Write a method ``MakeRectangle(width, height)`` that returns a
```
Comment (CarlyLanglois) [inline src/chapters/methods/exercises.rst]: this can be modified to be called 3. (right now it renders as 1. because of the text line breaking this up into 2 sections)

File: src/chapters/methods/exercises.rst [hunk 29]
```diff
@@ -0,0 +1,188 @@
...
+      #########
+      #########
+       #######
+        #####
+         ###
+          #
+
+   .. tip::
+
+      Consider what happens if you create a triangle and reverse it within the same method?
```
Comment (CarlyLanglois) [inline src/chapters/methods/exercises.rst:181]: im not getting the idea behind this tip. can you elaborate? or maybe just get rid of it?

File: src/chapters/methods/exercises.rst [hunk 30]
```diff
@@ -0,0 +1,188 @@
...
+
+   .. sourcecode:: csharp
+
+      MakeLine(5);
+
+   **Console Output**
+
+   ::
+
+      #####
```
Comment (CarlyLanglois) [inline src/chapters/methods/exercises.rst:61]: we should provide the signature for at least the first of these methods to get students into the groove. also though, if we're wanting these methods to return these shapes, then the "Console Output" will not be anything unless Console.WriteLine(MakeLine(5)) is called, right?

Maybe we should ask the students instead to write methods that return void but print the shapes instead? having the shapes be returned would require more practice with return types though

File: src/chapters/methods/studio.rst [hunk 31]
```diff
@@ -0,0 +1,88 @@
...
+Let's have some fun by creating a process that reverses BOTH the order of the
+entries in an array AND the order of characters within the individual elements.
+
+Remember that a method should perform only *one* task. 
+
+
+Warm-Up: Reverse a Number Array
+--------------------------------
+
+1. Define a method called ``ReverseNumbers`` that will sort the numbers within an array in descending order.
```
Comment (CarlyLanglois) [inline src/chapters/methods/studio.rst:59]: here's where you can say: create a new repl ... yada yada ... don't forget to create a container ``MainClass`` and copy the ``Main`` method signature so that you can run the program and call your ``ReverseNumbers`` method

File: src/chapters/methods/studio.rst [hunk 32]
```diff
@@ -0,0 +1,88 @@
+Studio: Methods
+====================
+
+The ``Reverse`` method flips the order of the elements within an array.
```
Comment (CarlyLanglois) [inline src/chapters/methods/studio.rst:13]: actually, having a note in the intro to the chapter on these built in array methods and how they have seen some methods before would be a great way to ease the reader into the new topic of what methods are. you can link back to the array methods appendix

File: src/chapters/methods/studio.rst [hunk 33]
```diff
@@ -0,0 +1,88 @@
...
+
+
+Warm-Up: Reverse a Number Array
+--------------------------------
+
+1. Define a method called ``ReverseNumbers`` that will sort the numbers within an array in descending order.
+2. Within the method, ``Sort`` the array.
+3. Use ``Reverse`` method to put the items into descending order.  
+4. Using a loop of your choice, print your final array to the console.
+5. Be sure to test your method, call it at least twice with arrays of differnt sizes and numbers.
```
Comment (CarlyLanglois) [inline src/chapters/methods/studio.rst]: different*

File: src/chapters/methods/studio.rst [hunk 34]
```diff
@@ -0,0 +1,88 @@
...
+4. Using a loop of your choice, print your final array to the console.
+5. Be sure to test your method, call it at least twice with arrays of differnt sizes and numbers.  
+
+
+Reverse a String
+------------------
+
+1. Define a method called ``ReverseLetters`` that will reverse each character inside a string and print the results to the console.  This will be similar to the ``ReverseNumbers`` method.
+2.  Look through the `String Methods <string-methods>`_  and `Array Methods <array-methods>`_ to figure out the best approach for this.  Consider the following: How can you access or alter characters within strings? What data types are going into your method, being used in your method, and printed by your method?
+3. Be sure to test your method by invoking it a few times.  Some test examples you could use: ``C Sharp``, ``Hello``, ``Coding is fun!`` they should return as ``prahS C``, ``olleH``, and ``!nuf so gniodC`` .
```
Comment (CarlyLanglois) [inline src/chapters/methods/studio.rst]: the last reverse would be : ``!nuf si gnidoC``

File: src/chapters/methods/studio.rst [hunk 35]
```diff
@@ -0,0 +1,88 @@
...
+4.  Now make this method take input from a user.  Print a message that contains the original string and the newly reversed string.
+   
+
+Palindrome Checker
+--------------------
+
+Now that you can completely reverse a string, let's use this to check to see if a word is a palindrome.
+A palindrome is a words that is spelled the same forwards and backwards.  Some examples are "tacocat", "kayak", and racecar".
```
Comment (CarlyLanglois) [inline src/chapters/methods/studio.rst]: and "racecar" - it's missing the opening quotation mark

File: src/chapters/methods/studio.rst [hunk 36]
```diff
@@ -0,0 +1,88 @@
...
+.. admonition:: Note
+
+   There are other factors that are sometimes included in the definition of a palindrome. For example, an alternative definition is that a palindrome is a sentence or phrase that contains letters in the same order, whether considered from beginning-to-end, or end-to beginning, ignoring punctuation, case, and spaces.
+
+One way to state a palindrome condition is to say that a palindrome is a string that is equal to its reverse.  
+When we think about it that way, what we need to do is create a method that will take a string, reverse it, compare it to the original and verify if they are equal.
+
+
+You have alredy built a function that reverses a string.  You could extend that method to be able to detect palindromes. 
+Or since, this data type for this will always be a string, you could build it using only string methods and a loop.
```
Comment (CarlyLanglois) [inline src/chapters/methods/studio.rst]: "Or since, this data type for this will always be a string," this construction is throwing me off.
Can you say instead: "Or you could build it using only string methods and a loop." Do we lose any critical info like this?

File: src/chapters/methods/studio.rst [hunk 37]
```diff
@@ -0,0 +1,88 @@
...
+
+One way to state a palindrome condition is to say that a palindrome is a string that is equal to its reverse.  
+When we think about it that way, what we need to do is create a method that will take a string, reverse it, compare it to the original and verify if they are equal.
+
+
+You have alredy built a function that reverses a string.  You could extend that method to be able to detect palindromes. 
+Or since, this data type for this will always be a string, you could build it using only string methods and a loop. 
+The choice is up to you.  
+
+1.  Define your method ``IsPalindrome``, to check if a string is equal to its reverse.  Think what data type the overall **outcome** will require.
```
Comment (CarlyLanglois) [inline src/chapters/methods/studio.rst]: i'd keep outcome as plain text (not bold)

**Review by CarlyLanglois — APPROVED** [review]
Overall really great content here Courtney! The notes we talked about this morning I think can mainly be addressed in the introductory page. Let's add some language around these ideas:

a) methods vs functions: id say, lets tell the students that technically c# programmers use only methods because methods MUST be defined within a class. but at times in the book and in conversations, its ok to interchange methods and functions
b) calling/invoking - invoking is defined on one of the first pages but you use the phrase 'call a method' a few times before its introduced or defined at all. can you add a quick definition of what it means to call a method. something to distinguish between the act of writing and defining the method and then using it/running it
c) the static/instance method definitions are a little off. from what we talked about this morning - can you do some slight refactoring of these pages to instead simply let the students know that methods have 'signatures' and we'll expand upon what a method signature requires in the classes chapters?

otherwise - the studio is an appropriate amount of work yes :)

File: src/chapters/methods/method-scope.rst (hunk 19 above)
Comment (speudusa) [reply src/chapters/methods/method-scope.rst]: i was matching replit, but VS does put the bracket below like the examples in unit 2.

File: src/chapters/methods/exercises.rst (hunk 30 above)
Comment (speudusa) [reply src/chapters/methods/exercises.rst:61]: does this mean we should give them the first method?  or the method template?  

i used void methods.

File: src/chapters/methods/exercises.rst (hunk 29 above)
Comment (speudusa) [reply src/chapters/methods/exercises.rst:181]: This was from the JS book.  I didn't really use it when working on solutions.  I left it in case someone found it helpful.  I;m not very attached to it, so I have no issue pulling it.

**Comment by speudusa** [comment]
made corrections based on the revisions provided.  Changed formatting of example presentation, language.  Created new chapter for basic class and object intro.  Tacked the Math Class on the end.  Didn't do any exercises or studio with this.  Classes will be easier in VS, and we'll get a lot of practice in the Class chapters.

I updated some of the Check Your Understanding questions.  Once this gets the go ahead to merge, i'll update Canvas with the new questions.

File: src/chapters/methods/what-is-a-method.rst [hunk 38]
```diff
@@ -0,0 +1,123 @@
...
+
+.. index:: ! method, ! function
+
+A **method** is a reusable, callable, and customizable piece of code. Methods are created to perform a single task or action.
+
+.. index:: ! method
+
+.. admonition:: Note
+
+   In C# or other OOP languages, developers use methods because they are assiociated with a class.
```
Comment (CarlyLanglois) [inline src/chapters/methods/what-is-a-method.rst]: OOP is not yet defined - otherwise great note!

File: src/chapters/methods/method-signatures.rst [hunk 39]
```diff
@@ -0,0 +1,211 @@
...
+      {
+         return num * num;
+      }
+
+
+In this example, we are skipping the **access modifier**.  
+The reason for this is that we are working within a single class.  
+If we had multiple classes this would be more significant.  We will discuss this more when we get to class and class members.
+
+We are using the ``static`` modifier.   This is lets C# know that this method
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-signatures.rst]: This lets C#* ...

File: src/chapters/methods/method-signatures.rst [hunk 40]
```diff
@@ -0,0 +1,211 @@
...
+Check Your Understanding
+------------------------
+.. admonition:: Question
+
+   What is the expected output returned by the ``Doubles`` method if the input, ``x``, was 7?
+
+   .. sourcecode:: csharp
+      :linenos:
+         
+         static int Doubles(int x)
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-signatures.rst]: i think the indentation on this codeblock is off

File: src/chapters/methods/using-methods.rst [hunk 41]
```diff
@@ -0,0 +1,302 @@
...
+   **Console Output**
+
+   :: 
+      
+      Willow's birthday is on 12/12
+      Lil's birthday is on 8/21
+      
+
+In this example, we are asking the method to print a friend's name and birth date inside a preprogrammed message.  
+This might see really simple, but if you had a list of names to run through, calling your method for each is a LOT less typing than
```
Comment (CarlyLanglois) [inline src/chapters/methods/using-methods.rst]: This might seem*

File: src/chapters/methods/method-scope.rst [hunk 42]
```diff
@@ -0,0 +1,194 @@
...
+
+.. admonition:: Example
+
+   .. sourcecode:: csharp
+      :linenos:
+
+      class MainClass 
+      {
+
+      static void PrintThis()
```
Comment (CarlyLanglois) [inline src/chapters/methods/method-scope.rst]: check the indentation levels here too

File: src/chapters/classes-objects-math/class-type.rst [hunk 43]
```diff
@@ -0,0 +1,116 @@
...
+Now that you have a better understanding of methods, we should touch on classes and objects *briefly*.  
+These are big concepts in C#, and OOP in general.  
+This chapter is designed to help you get familiar with the larger concepts of classes and objects.
+We will explore definitions and walkthrough examples that will demonstarate how classes, objects, and methods work together.
+
+
+Class Type
+-------------------
+
+Quick refesher on data types.  As discussed in chapter 4, C# contains value types and reference types.
```
Comment (CarlyLanglois) [inline src/chapters/classes-objects-math/class-type.rst]: can you replace the reference to chapter 4 with a reference to the section title - and you can even use a :ref: link to give a hyperlink back to that page.
we want to avoid referring to chapter numbers because if we move sections around, we dont want to have to check the book for references like this

File: src/chapters/classes-objects-math/classes-objects-methods.rst [hunk 44]
```diff
@@ -0,0 +1,363 @@
...
+The template dictates what an object of this class requires, as well as how that object behaves.
+
+To declare a new class, you will need the keyword ``class``.  
+
+.. admonition:: Note
+
+   When working with replit.com, you will need to create a new class in a new file.  
+   In order for your class to be recognized by ``MainClass`` in replit, you will need to provide your class its own unique ``namespace``.
+   
+   We will be moving to a differnt IDE soon, so any replit.com code examples will be provided to you.
```
Comment (CarlyLanglois) [inline src/chapters/classes-objects-math/classes-objects-methods.rst]: different*

File: src/chapters/classes-objects-math/classes-objects-methods.rst [hunk 45]
```diff
@@ -0,0 +1,363 @@
...
+
+.. admonition:: Note
+
+   When working with replit.com, you will need to create a new class in a new file.  
+   In order for your class to be recognized by ``MainClass`` in replit, you will need to provide your class its own unique ``namespace``.
+   
+   We will be moving to a differnt IDE soon, so any replit.com code examples will be provided to you.  
+
+When working in a project with multiple classes, it is common to have a ``namespace`` that connects all classes within a project.
+In this context, a ``namespace`` can be more for organizational purposes, as in it will provide direction to each class within
```
Comment (CarlyLanglois) [inline src/chapters/classes-objects-math/classes-objects-methods.rst]: Can you convert this to two sentences?

File: src/chapters/classes-objects-math/classes-objects-methods.rst [hunk 46]
```diff
@@ -0,0 +1,363 @@
...
+
+   When working with replit.com, you will need to create a new class in a new file.  
+   In order for your class to be recognized by ``MainClass`` in replit, you will need to provide your class its own unique ``namespace``.
+   
+   We will be moving to a differnt IDE soon, so any replit.com code examples will be provided to you.  
+
+When working in a project with multiple classes, it is common to have a ``namespace`` that connects all classes within a project.
+In this context, a ``namespace`` can be more for organizational purposes, as in it will provide direction to each class within 
+the main project.  If the ``namespace`` is not part of the class that holds your ``Main method`` you will have to access it through a ``using`` statement.
+Just like we did will Collections.  This tells the compiler to look through any classes within the same ``namespace`` for definitions and requirements.
```
Comment (CarlyLanglois) [inline src/chapters/classes-objects-math/classes-objects-methods.rst]: with* Collections

File: src/chapters/classes-objects-math/classes-objects-methods.rst [hunk 47]
```diff
@@ -0,0 +1,363 @@
...
+ //data type                  constructor with possilbe parameters
+   ClassName objectName = new ClassName(possible parameters); 
+
+
+In our ``Cat`` class example, we instantiate ``myCat`` which is a new ``Cat`` object.
+
+.. sourcecode:: csharp
+   :lineno-start: 10
+
+      Cat myCat = new Cat();
```
Comment (CarlyLanglois) [inline src/chapters/classes-objects-math/classes-objects-methods.rst]: more indentations to check here

File: src/chapters/classes-objects-math/classes-objects-methods.rst [hunk 48]
```diff
@@ -0,0 +1,363 @@
...
+   :lineno-start: 10
+
+      Cat myCat = new Cat();         
+
+
+Class Constructors
+^^^^^^^^^^^^^^^^^^^^^^^
+
+When we instantiate an object, we use the class **constructor** to provide any needed values to initialize any attributes.  
+The syntax just like the parameter list of a method.
```
Comment (CarlyLanglois) [inline src/chapters/classes-objects-math/classes-objects-methods.rst]: this is not a complete sentence oups! not sure what you're conveying here

File: src/chapters/classes-objects-math/classes-objects-methods.rst [hunk 49]
```diff
@@ -0,0 +1,363 @@
...
+   Cat myCat = new Cat();
+
+
+Class Members
+^^^^^^^^^^^^^^^
+
+Class definitions deterimine any **attributes** the object requires, as well as any **behaviors** of the object.
+Attributes are class variables, also called **fields**.  They can be hard coded, or filled in at the time of instantiation.
+
+Methods are the behvaiors or actions your class is able to preform on objects created by that class.
```
Comment (CarlyLanglois) [inline src/chapters/classes-objects-math/classes-objects-methods.rst]: behaviors* and perform*

File: src/chapters/classes-objects-math/classes-objects-methods.rst [hunk 50]
```diff
@@ -0,0 +1,363 @@
...
+**Console Output**
+
+.. sourcecode:: csharp
+
+   Alyce
+   Purr
+
+
+Since we used the default constructor to create the ``myCat`` object, 
+the attribute ``name`` will "Alyce" or exactly what we coded it to be.
```
Comment (CarlyLanglois) [inline src/chapters/classes-objects-math/classes-objects-methods.rst]: I think you're missing a verb in this sentence - will be*?

File: src/chapters/classes-objects-math/math-class.rst [hunk 51]
```diff
@@ -0,0 +1,159 @@
...
+
+.. index:: ! const, ! constants ! PI ! pi ! Pi
+
+``Math`` Fields Are Constants
+------------------------------
+
+Having a built in ``Math`` class can be very helpful when it comes to common math elements, such as ``PI``.  
+
+One of the benefits of constant values is that you don't have to memorize them, or define them in a variable.
+C# stores the value of pi upto 16 decimal places for us.
```
Comment (CarlyLanglois) [inline src/chapters/classes-objects-math/math-class.rst]: up to - needs to be two words

**Review by CarlyLanglois — APPROVED** [review]
excellent additions! The methods revisions are really well done - thanks for going back and making those changes. 
I'm digging this classes and objects introduction as well - I suspect we may have some overlap with the classes pt 1 section but we'll return to that talk when you move those pages over i suppose.
